src/libs/kata-sys-util/src/fs.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::ffi::OsString;
+use std::fs::{self, File};
+use std::io::{Error, Result};
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use crate::{eother, sl};
+
+// from linux.git/fs/fuse/inode.c: #define FUSE_SUPER_MAGIC 0x65735546
+const FUSE_SUPER_MAGIC: u32 = 0x65735546;
+
+/// Get bundle path (current working directory).
+pub fn get_bundle_path() -> Result<PathBuf> {
+    std::env::current_dir()
+}
+
+/// Get the basename of the canonicalized path
+pub fn get_base_name<P: AsRef<Path>>(src: P) -> Result<OsString> {
+    let s = src.as_ref().canonicalize()?;
+    s.file_name().map(|v| v.to_os_string()).ok_or_else(|| {
+        eother!(
+            "failed to get base name of path {}",
+            src.as_ref().to_string_lossy()
+        )
+    })
+}
+
+/// Check whether `path` is on a fuse filesystem.
+pub fn is_fuse_fs<P: AsRef<Path>>(path: P) -> bool {
+    if let Ok(st) = nix::sys::statfs::statfs(path.as_ref()) {
+        if st.filesystem_type().0 == FUSE_SUPER_MAGIC as i64 {
+            return true;
+        }
+    }
+    false
+}
+
+/// Check whether `path` is on a overlay filesystem.
+pub fn is_overlay_fs<P: AsRef<Path>>(path: P) -> bool {
+    if let Ok(st) = nix::sys::statfs::statfs(path.as_ref()) {
+        if st.filesystem_type() == nix::sys::statfs::OVERLAYFS_SUPER_MAGIC {
+            return true;
+        }
+    }
+    false
+}
+
+/// Check whether the given path is a symlink.
+pub fn is_symlink<P: AsRef<Path>>(path: P) -> std::io::Result<bool> {
+    let path = path.as_ref();
+    let meta = fs::symlink_metadata(path)?;
+
+    Ok(meta.file_type().is_symlink())
+}
+
+/// Reflink copy src to dst, and falls back to regular copy if reflink copy fails.
+///
+/// # Safety
+/// The `reflink_copy()` doesn't preserve permission/security context for the copied file,
+/// so caller needs to take care of it.
+pub fn reflink_copy<S: AsRef<Path>, D: AsRef<Path>>(src: S, dst: D) -> Result<()> {
+    let src_path = src.as_ref();
+    let dst_path = dst.as_ref();
```

Comment:
A few security-related questions:

- Can `src` and `dst` be _any_ values?
- Should they be absolute paths?
- Should either path be below a known sub-directory?
- Do we care about the `src` owner and the `dst` directory owner?

Reply:
This is an executor, it doesn't have information about the context, so policy should be adopted at high layer.